### PR TITLE
fix(ingest-idp): emit empty GroupMembership when there are no groups

### PR DESCRIPTION
**Pull request by aditya-radhakrishnan**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/identity/okta.py [hunk 1]
```diff
@@ -316,7 +317,9 @@ def get_workunits(self) -> Iterable[MetadataWorkUnit]:
                 yield group_status_wu
 
         # Step 2: Populate GroupMembership Aspects for CorpUsers
-        datahub_corp_user_urn_to_group_membership: Dict[str, GroupMembershipClass] = {}
+        datahub_corp_user_urn_to_group_membership: Dict[
+            str, GroupMembershipClass
+        ] = defaultdict(lambda: GroupMembershipClass(groups=[]))
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/identity/okta.py:322]: The logic here is pretty complicated so I might be wrong, but I think this requires some changes below too. defaultdict makes dict[key] always succeed, but `key in dict` will still return false if the key hasn't been written to

**Review by hsheth2 — CHANGES_REQUESTED** [review]
Looks like the tests are failing

File: metadata-ingestion/src/datahub/ingestion/source/identity/okta.py (hunk 1 above)
Comment (aditya-radhakrishnan) [reply metadata-ingestion/src/datahub/ingestion/source/identity/okta.py:322]: Should be fixed!

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 2]
```diff
@@ -419,18 +419,10 @@ def ingest_ad_users(
...
-                datahub_group_membership = (
-                    datahub_corp_user_urn_to_group_membership.get(
-                        datahub_corp_user_snapshot.urn
-                    )
-                )
-                assert datahub_group_membership
-                datahub_corp_user_snapshot.aspects.append(datahub_group_membership)
+            datahub_corp_user_urn: str = datahub_corp_user_snapshot.urn
+            datahub_group_membership: GroupMembershipClass = datahub_corp_user_urn_to_group_membership[datahub_corp_user_urn]
+            assert datahub_group_membership is not None
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:424]: I think this assert is not required

also the "Add GroupMembership" comment was still useful

**Comment by jjoyce0510** [comment]
(Tests still failing)

File: metadata-ingestion/src/datahub/ingestion/source/identity/okta.py [hunk 3]
```diff
@@ -364,18 +357,11 @@ def get_workunits(self) -> Iterable[MetadataWorkUnit]:
...
-                    datahub_group_membership = (
-                        datahub_corp_user_urn_to_group_membership.get(
-                            datahub_corp_user_snapshot.urn
-                        )
-                    )
-                    assert datahub_group_membership is not None
-                    datahub_corp_user_snapshot.aspects.append(datahub_group_membership)
+                # Add GroupMembership aspect populated in Step 2.
+                datahub_corp_user_urn: str = datahub_corp_user_snapshot.urn
+                datahub_group_membership: GroupMembershipClass = datahub_corp_user_urn_to_group_membership[datahub_corp_user_urn]
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/identity/okta.py:362]: this code looks eerily similar to the azure code above. we probably want to refactor it in the future

File: metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py [hunk 4]
```diff
@@ -107,6 +107,61 @@ def test_azure_ad_source_default_configs(pytestconfig, tmp_path):
...
+                        "client_secret": "client_secret",
+                        "redirect": "https://login.microsoftonline.com/common/oauth2/nativeclient",
+                        "authority": "https://login.microsoftonline.com/00000000-0000-0000-0000-000000000000",
+                        "token_url": "https://login.microsoftonline.com/00000000-0000-0000-0000-000000000000/oauth2/token",
+                        "graph_url": "https://graph.microsoft.com/v1.0",
+                        "ingest_group_membership": True,
+                        "ingest_groups": True,
+                        "ingest_users": True,
+                    },
+                },
```
Comment (hsheth2) [inline metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py:146]: if possible, could we use a pytest parameterized test

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py (hunk 2 above)
Comment (aditya-radhakrishnan) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:424]: Can do!

File: metadata-ingestion/src/datahub/ingestion/source/identity/okta.py (hunk 3 above)
Comment (aditya-radhakrishnan) [reply metadata-ingestion/src/datahub/ingestion/source/identity/okta.py:362]: Yeah I know...I'll write a TODO here

File: metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py (hunk 4 above)
Comment (aditya-radhakrishnan) [reply metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py:146]: wdym?

Comment (aditya-radhakrishnan) [reply metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py:146]: or do you have an example i can look at?

**Review by jjoyce0510 — APPROVED** [review]

**Pull request merged**